lib/java/persistance/src/main/java/org/enso/persist/PerGenerator.java

```diff
@@ -106,6 +114,63 @@ final int versionStamp() {
     return map.versionStamp;
   }
 
+  private int registerReference(Persistance.Reference<?> ref) {
+    var obj = ref.get(Object.class);
+    var existingId = pendingReferences.get(obj);
+    if (existingId == null) {
+      var currentSize = countReferences++;
+      pendingReferences.put(obj, currentSize);
+      return currentSize;
+    } else {
+      return existingId;
+    }
+  }
+
+  /**
+   * Writes an object into the buffer. Writes also all {@link Persistance.Reference} that were left
+   * pending during the serialization.
+   *
+   * @param obj the object to write down
+   * @return location of the table {@code int size and then int[size]}
+   */
+  private int writeObjectAndReferences(Object obj) throws IOException {
+    var objAt = writeObject(obj);
+
+    var refsOut = new ByteArrayOutputStream();
+    var refsData = new DataOutputStream(refsOut);
+    refsData.writeInt(-1); // space for size of references
+    refsData.writeInt(objAt); // the main object
+    var count = 1;
+    while (!pendingReferences.isEmpty()) {
+      var round = new ArrayList<>(pendingReferences.entrySet());
+      round.sort(
+          (e1, e2) -> {
+            return e1.getValue() - e2.getValue();
+          });
+      pendingReferences.clear();
+      for (var entry : round) {
+        count++;
+        if (obj == entry.getKey()) {
+          refsData.writeInt(-1);
+          continue;
+        }
+        var at = writeObject(entry.getKey());
+        assert count == entry.getValue() : "Expecting " + count + " got " + entry.getValue();
```

Comment:
tbh I don't think I understand what guarantees that in every iteration this assertion is true.

Reply:
Hmm, is it a general assumption that the local variable `count` and the field `countReferences` are kept in-sync (i.e. the `countReferences` has the same value when the reference is first created as `count` has when it is written)?

Reply:
But to my understanding that would suggest that `writeObjectAndReferences` can only be called once for each instance. Is my assumption correct? If so, can we enforce it? (e.g. `if (countReferences > 0) { throw ... }` at the start? Or if not correct, what mistake in understanding am I making?

Reply:
It is a private method. Nobody outside of this class can call it multiple times and the class does it exactly once.